### DISCUSSION TITLE
Moving definitions depending on sixtyfps to other file behaves weird

### DISCUSSION THREAD

**Discussion by DaMilyutin** [slint-ui/slint#778] [Q&A] [answered]
I've extracted some helper struct to other rs file and get error like

"

error[E0271]: type mismatch resolving `<VecModel<id_map_data::sixtyfps_generated_AppWindow::ListItemData> as Model>::Data == sixtyfps_generated_AppWindow::ListItemData`
  --> src\main.rs:19:9
   |
19 |         task_data.as_ref().borrow().data.clone(),
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected struct `sixtyfps_generated_AppWindow::ListItemData`, found struct `id_map_data::sixtyfps_generated_AppWindow::ListItemData`
   |
   = note: required for the cast to the object type `dyn Model<Data = sixtyfps_generated_AppWindow::ListItemData>`
"
It compiles and works when I copy all stuff in main.rs.

So is it possible to move some sixtyfps stuff to other rs files?

**Comment by Vollbrecht** [comment]
If i understand you correctly your goal is to get the struct data declared in sixtyfps in your rust code, potential modify it and set it back? Because the sixtyfps model trait is implemented multiple times depending on the trait bounds you have to understand how you try to access the model data. To have a look at all the implementations have a look at the [model source file](https://github.com/sixtyfpsui/sixtyfps/blob/master/sixtyfps_runtime/corelib/model.rs) 
If your ListItemData is an defined array in sixtyfps you could access and modify its content with the following pseudo code
```
let the_answer = 42.0;
let mut my_count = ui.get_ListItemData().row_count();
for i in 0..my_count {
    let mut foo = ui.get_ListItemDatat().row_data(i);
    foo.some_data = the_answer;
    ui.get_ListItemData().set_row_data(i, foo);
}
```
hope that helps

**Comment by DaMilyutin** [comment, marked as answer]
@Vollbrecht , Hi. I guess best way to explain is to show code:
[extract-to-module-not-working-for-me.zip](https://github.com/sixtyfpsui/sixtyfps/files/7794971/extract-to-module-not-working-for-me.zip)
 
When I cut and paste my code from id_map_data.rs to main.rs it works.
 
[when-it-works.zip](https://github.com/sixtyfpsui/sixtyfps/files/7794980/when-it-works.zip)

I assume it is because types declared in different files are somehow considered/generated as different.
Maybe, I use macro `sixtyfps::include_modules!();` wrong?

**Reply by tronical** [reply]
Hi @DaMilyutin , you can use `sixtyfps::include_modules!();` only once per crate. In your specific case you can, for example, in `id_map_data.rs` replace the use of the macro with `use super::*;`. Or you can use the macro in `main.rs` within a module, to encapsulate further.
